File: backend/models/roadmap.py
from pydantic import BaseModel
from typing import List, Optional
from .spec import UserSpec

class Resource(BaseModel):
    """
    Represents a specific educational resource (link) for a topic.
    Types: "docs", "video", "article", "interactive"
    """
    label: str
    url: str
    type: str

class Topic(BaseModel):
    """
    A specific unit of study within a phase.
    Contains content summary, sub-tasks, and curated resources.
    """
    name: str
    content: Optional[str] = ""
    subtopics: List[str]
    resources: List[Resource] = []
    project: Optional[str] = None

class Phase(BaseModel):
    """
    A logical group of topics, typically spanning a specific timeframe (e.g., Week 1-2).
    """
    phase_number: int
    title: str
    week_range: str
    topics: List[Topic]

class Roadmap(BaseModel):
    """
    The root data model for a complete learning roadmap.
    Statuses: "pending", "generating", "complete", "error"
    """
    id: str
    title: str
    spec: UserSpec
    phases: List[Phase]
    created_at: str
    status: str = "pending"

File: frontend/src/types/roadmap.ts
export type SkillLevel = 'beginner' | 'intermediate' | 'advanced'
export type TopicStatus = 'not_started' | 'in_progress' | 'done'
export type AgentName = 'analyst' | 'curriculum' | 'resources' | 'formatter'
export type SSEEventType = 'agent_start' | 'agent_done' | 'agent_error' | 'complete' | 'error'

export interface Resource {
  label: string
  url: string
  type: string // "docs" | "video" | "article" | "interactive"
}

export interface Topic {
  name: string
  subtopics: string[]
  resources: Resource[]
  project?: string
}

export interface Phase {
  phase_number: number
  title: string
  week_range: string
  topics: Topic[]
}

export interface UserSpec {
  goal: string
  skill_level: SkillLevel
  hours_per_week: number
  estimated_weeks: number
}

export interface Roadmap {
  id: string
  title: string
  spec: UserSpec
  phases: Phase[]
  created_at: string
  status: string // "pending" | "generating" | "complete" | "error"
}

export interface ProgressState {
  roadmap_id: string
  topics: Record<string, TopicStatus>
}

export interface SSEEvent {
  type: SSEEventType
  agent?: AgentName
  roadmap_id?: string
  message?: string
}

export interface RoadmapListItem {
  id: string
  title: string
  created_at: string
  status: string
  topic_count: number
}
